server/src/internalClusterTest/java/org/elasticsearch/discovery/DiscoveryDisruptionIT.java

```diff
@@ -257,4 +260,60 @@ public void testJoinWaitsForClusterApplier() throws Exception {
         logger.info("--> waiting for cluster to heal");
         ensureStableCluster(3);
     }
+
+    public void testJoinWaitsForCircuitBreaker() throws InterruptedException {
+        startCluster(3);
+
+        final var masterName = internalCluster().getMasterName();
+        final var victimName = randomValueOtherThan(masterName, () -> randomFrom(internalCluster().getNodeNames()));
+        logger.info("--> master [{}], victim [{}]", masterName, victimName);
+
+        // fill up the circuit breaker to breaking point
+        final var circuitBreaker = internalCluster().getInstance(CircuitBreakerService.class, victimName)
+            .getBreaker(CircuitBreaker.IN_FLIGHT_REQUESTS);
+        long allocationSize = 1;
+        while (true) {
+            try {
+                circuitBreaker.addEstimateBytesAndMaybeBreak(allocationSize, "test");
+            } catch (CircuitBreakingException e) {
+                circuitBreaker.addWithoutBreaking(allocationSize);
+                break;
+            }
+            allocationSize <<= 1;
+            assert 0 <= allocationSize;
+        }
+
+        // drop the victim from the cluster with a network disruption
+        final var masterTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class, masterName);
+        masterTransportService.addFailToSendNoConnectRule(internalCluster().getInstance(TransportService.class, victimName));
+        logger.info("--> waiting for victim's departure");
+        ensureStableCluster(2, masterName);
+
+        // verify that the victim sends no joins while the circuit breaker is breaking
+        final var victimTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class, victimName);
+        victimTransportService.addSendBehavior((connection, requestId, action, request, options) -> {
+            assertNotEquals(action, JoinHelper.JOIN_ACTION_NAME);
+            connection.sendRequest(requestId, action, request, options);
+        });
+
+        // fix the network disruption
+        logger.info("--> removing network disruption");
+        masterTransportService.clearAllRules();
+        ensureStableCluster(2, masterName);
```

Comment:
I am surprised this is still 2 nodes. Should this permit to connect the node again?

Reply:
No, the victim node still can't join the cluster here because it has a tripping circuit breaker (that's the point!). We have to release the resources tracked by the CB to get it to join again.